### ISSUE TITLE
Overtraining the model

### ISSUE THREAD

**Issue by maciejdomagala** [jdb78/pytorch-forecasting#242]
Hi!
 
I have been experimenting with the TFT model for quite a bit and noticed that I cannot get the model to _perfectly overfit_ on my training data. What I mean by that, I would just like the network to memorize correct future sequences of prices for _most of_ the training dataset (validation and testing are not important here).

About the data, I am working with stock data for just one financial index, data in 3 years span (~700 rows of data). For simplicity, I am only processing 4 features (open, high, low, close) with the _close_ price being the target of the prediction, also 'month' as categorical feature. I am using a window of length 20 as input on the encoder (min_encoder_length=max_encoder_length=20) and my future range is 5 (min_prediction_length=max_prediction_length=5). I am using EncoderNormalizer for the target data (changed in the source code so it works for sequence of length 20 also, didn't like the idea of leaking any information). That would be it for the data settings, I can include all of the settings if that would be more helpful.

As for the model, I went with default output_size=7 and QuantileLoss. I increased the capacity of the model by setting hidden_size=256 (I tried 512 also) and setting 4 attention heads and 4 lstm layers. It results in nearly 7M params in the model (2M on the LSTM encoder), should be way more than enough for the amount of data that I have. I set the dropout to be some low value (0.001% or so) and went with the 1e-3 learning rate with the default Ranger. 
I do not have the EarlyStopping callback enabled.

The train_loss hits the plateau at around 0.002 loss value after ~60 epochs. (batch steps on the image below).

![image](https://user-images.githubusercontent.com/54290898/103633079-835a4100-4f45-11eb-9ec1-168d23a6c085.png)

Random training predictions that I made are acting quite _linear_ in most cases and are in a similar range as targets, but are not following the exact paths (which I would expect with heavy overfitting).

![image](https://user-images.githubusercontent.com/54290898/103633433-011e4c80-4f46-11eb-8fff-20f778eadd0e.png)
![image](https://user-images.githubusercontent.com/54290898/103633530-21e6a200-4f46-11eb-848e-db9bd3057c1f.png)
![image](https://user-images.githubusercontent.com/54290898/103633566-33c84500-4f46-11eb-88b3-000987b33e19.png)

My questions would be - is it correct for me to assume that I can overfit with almost 100% precision on the training data or am I misunderstanding the concept? Are there some obvious flaws with the training setup that I described above?

It is very likely that I am not using the framework in the way it should be used - I will gladly take any advice. Hope this is a good place for this kind of issue, if not - let me know.

**Comment by jdb78** [comment]
Can you please post the definition of the `TimeSeriesDataSet`?

**Comment by maciejdomagala** [comment]
```
dataset_settings = dict(time_idx="time_idx", target="close",
                        group_ids=["instrument_id"],
                        min_encoder_length=20,  # allow encoder lengths from 0 to max_prediction_length
                        max_encoder_length=20,
                        min_prediction_length=5,
                        max_prediction_length=5,
                        static_reals=[],
                        time_varying_known_categoricals=["month"],
                        variable_groups={},  # group of categorical variables can be treated as one variable
                        time_varying_known_reals=[],
                        time_varying_unknown_categoricals=[],
                        time_varying_unknown_reals=[
                            "open",
                            "high",
                            "low",
                            "close"
                        ],
                        )
```

These are my settings, I am calling ```TimeSeriesDataSet``` on these with my training dataframe.

**Comment by jdb78** [comment]
Can you add the `time_idx` as a known real variable? Potentially also use `add_relative_time_idx=True` and check if you still have issues. If you also have a large hidden_size of say 256, you should be able to overfit.

**Comment by maciejdomagala** [comment]
Added these into settings as you suggested
```
dataset_settings = dict(time_idx="time_idx", target="close",
                        group_ids=["instrument_id"],
                        min_encoder_length=20,  # allow encoder lengths from 0 to max_prediction_length
                        max_encoder_length=20,
                        min_prediction_length=5,
                        max_prediction_length=5,
                        static_reals=[],
                        time_varying_known_categoricals=["month"],
                        variable_groups={},  # group of categorical variables can be treated as one variable
                        time_varying_known_reals=['time_idx'],
                        time_varying_unknown_categoricals=[],
                        time_varying_unknown_reals=[
                            "open",
                            "high",
                            "low",
                            "close"
                        ],
                        add_relative_time_idx=True,
                        )
```
I ran the model twice with these and previous settings, it doesn't seem to make any impact on the training (metrics and training process seems to be exactly the same with some minor random differences).

![image](https://user-images.githubusercontent.com/54290898/103820877-5acd6680-506d-11eb-845e-4d993d89fc0d.png)

Prediction plots are the same as above, in the range of actual values but not overfitted.

These are my model settings, although I am not sure whether there is something unusual with the setup.

```
tft = TemporalFusionTransformer.from_dataset(
    train,
    learning_rate=0.001,
    attention_head_size=2,
    dropout=0.001,
    hidden_size=256,
    hidden_continuous_size=128,
    lstm_layers = 1,
    output_size=7,
    loss=QuantileLoss(),
    log_interval=10,
    log_val_interval=1,
    reduce_on_plateau_patience=self.patience,
    )
```
Would you have any other suggestions? I can provide more info if needed.

**Comment by jdb78** [comment]
Do you happen to have a colab notebook to play with?

**Comment by maciejdomagala** [comment]
Bunch of scripts actually, will try to get it all together into jupyter and post here, thanks for helping me!

**Comment by maciejdomagala** [comment]
Alright, it seems that after some debugging I found the cause of the problem.
If anyone is interested - it seems that it was the case of data normalization. I changed the source code so it used the EncoderNormalizer by default - I wanted to avoid look-ahead bias on the training dataset as well. It turned out that, regardless of model capacity and parameters/loss functions, it would hit the wall at some point not converging, I think plots are somewhat intuitive:

![Zaznaczenie_052](https://user-images.githubusercontent.com/54290898/104639897-38100300-56a8-11eb-9c46-089163e05503.png)
![Zaznaczenie_051](https://user-images.githubusercontent.com/54290898/104639914-3c3c2080-56a8-11eb-9c38-cd0b562d992d.png)
![Zaznaczenie_049](https://user-images.githubusercontent.com/54290898/104639947-44945b80-56a8-11eb-95db-8dddb54977c0.png)

After changing the standarization to GroupNormalizer, the model seems to be converging in the way one would expect:

![Zaznaczenie_046](https://user-images.githubusercontent.com/54290898/104640097-773e5400-56a8-11eb-92e0-6e74b45249a2.png)
![Zaznaczenie_047](https://user-images.githubusercontent.com/54290898/104640193-84f3d980-56a8-11eb-9111-c7b2f0a0d64c.png)
![Zaznaczenie_045](https://user-images.githubusercontent.com/54290898/104640000-52e27780-56a8-11eb-8ec6-ae6e55ff0c83.png)

These were calculated using RMSE as a loss function, loss is getting visibly lower during the training (can be seen on the plots too). 
I still didn't quite build the intuition on why exactly this is happening - my best guess would be that there are periods in my training data with vastly different value ranges than others, e.g. two weeks with close price value 1.0-1.05 and month after that values like 1.30-1.35. If i understand this transform correctly, the GroupNormalizer handles the data in a sklearn.StandardScaler manner, dealing with it globally. On the other hand EncoderNormalizer might go crazy since it works only with a small window on a different value scale and the time.
Problem is solved though, I was able to overtrain, maybe you will find it somewhat useful. Cheers!

**Comment by quincy-kh-chen** [comment]
Thank you @jdb78 for this awesome repository and @maciejdomagala for your insights in this thread!

I'm also currently setting up an overfitting training using TFT model to fit on similar time series, mostly refactored from the [tutorial](https://pytorch-forecasting.readthedocs.io/en/latest/tutorials/stallion.html#Predict-on-selected-data). However, it looks like the overfitting was not successful. @maciejdomagala would it be possible if you can share the final parameters you used (training script if possible)? Such as:

- `target_normalizer`, `scalers`, `time_varying_unknown_reals`, `add_relative_time_idx`, `add_target_scales`  for TimeSeriesDataSet
- `loss`, `dropout` for TemporalFusionTransformer

Appreciate!
